woocommerce/google-listings-and-ads
# Scheduled notifications using meta handler

### Changes proposed in this Pull Request:

This PR adds some work for scheduling the Notifications in batches. 
The strategy of this PR is to set up some product metas and then filter them in a Notification Job.

For that this PR:

- Add single Product Notifcation Job for Create, Delete, Update 
- Adapt `ProductHelper.php` with some helper functions regarding Notifications.
- Adapt `ProductMetaHandler.php` creating some metas for the notifications.
- Create `NotificationStatus.php` to hold the Statuses for the Notification metas
- Tweak `ProductRepository.php` adding some functions to filter the products by Notifications meta.
- Update `SyncerHooks.php` in order to trigger the Notification Jobs.
- Tweak `NotificationsService.php` adding some functions and dependencies.

The way this works is that on Update Product we check if the Notifications are enabled. (For now using a filter). If so, instead of sync to the Merchant we updated some product metas marking a product as PENDING_CREATE, PENDING_DELETE or PENDING_UPDATE. Then we schedule the Jobs that filter those products based on the metas in batches. Finally, we process those items and update their metas to CREATED, UPDATED, DELETED.

The advantages of this strategy are:

- It impacts less in other jobs as it doesn't need to tweak the Action Scheduler Abstract Classes. 
- More efficient and clean in terms of code and collisions. For example, if a product gets pending update in the middle of a schedule it can be notified in the next schedule
- Can be easily refactored to use recurrent jobs and be even more efficient


### Detailed test instructions:


No need to test in detail the PR. Just offering me some feedback about the strategy. Final PR will come afterwards.

Here are some test to see how it works

1. Create a Product
2. See in Action Scheduler a new job created --> `gla/jobs/notifications//process_item` with one item (the product ID just created) and product.created as topic
3. Run this action
4. Go to Woocommerce - Status - Logs and see the Notify debug Log created  with the product.create topic
5. Update the product 
6. See in Action Scheduler a new job created --> `gla/jobs/notifications/process_item` with one item (the product ID just created) and product.update as topic
7.  Run this action
8. Go to Woocommerce - Status - Logs and see the Notify debug Log created with the product.update topic 
9. Send the product to trash
10. 6. See in Action Scheduler a new job created --> `gla/jobs/notifications/process_item` with one item (the product ID just created) and product.delete as topic
11. Run the action
12. Go to Woocommerce - Status - Logs and see the Notify debug Log created  with the product.delete topic
13. Untrash the product
14. See in Action Scheduler a new job created --> `gla/jobs/notifications/process_item` with one item (the product ID just restored) and product.create as topic
15. Run this action
16. Go to Woocommerce - Status - Logs and see the Notify debug Log created  with the product.create topic
17. Set Dont sync and show in the product
18. See in Action Scheduler a new job created --> `gla/jobs/notifications/process_item` with one item (the product ID just restored) and product.delete as topic
19. Run this action
20. Go to Woocommerce - Status - Logs and see the Notify debug Log created  with the product.delete topic
21. Update the product
22. Nothing is Scheduled (because is set as don't sync and show)
23. Set Sync and Show
24. See in Action Scheduler a new job created --> `gla/jobs/notifications/process_item` with one item (the product ID just restored) and product.create as topic
25. Run this action
26. Go to Woocommerce - Status - Logs and see the Notify debug Log created  with the product.create topic
27. Try to update several times to see there are not several updates happening 
28. Save the product as draft
29. See in Action Scheduler a new job created --> `gla/jobs/notifications/process_item` with one item (the product ID just restored) and product.delete as topic
30. Run the action
31. Go to Woocommerce - Status - Logs and see the Notify debug Log created  with the product.delete topic
32. Save the product as publish
33. See in Action Scheduler a new job created --> `gla/jobs/notifications//process_item` with one item (the product ID just created) and product.created as topic




### Additional details:


### Changelog entry

>
